### ISSUE TITLE
Should a service worker be allowed to register new service workers?

### ISSUE THREAD

**Issue by mkruisselbrink** [w3c/ServiceWorker#1117]
In the spec as currently written `navigator.serviceWorker.register` is exposed in all worker contexts, including service workers. That will then allow a service worker to register arbitrarily many new service workers (which could be nicely abused for some sort of nested worker support), as long as it uses different scopes for each. Is that desirable though?
Currently no implementation seems to actually have implemented `navigator.serviceWorker` in service worker contexts (or in any worker contexts for that matter, as far as I can tell). But if they do, and this is actually something that should be supported implementations need to be careful that this isn't an avenue that can be used to keep a worker alive indefinitely (by "forking" itself repeatedly before it is killed).

**Comment by waxter013** [comment]
It would be a **wasted opportunity** if one service worker could not register another. The goal of a service worker is to allow scripts to respond to events in the background. Some events will require specific scripts to handle them. 

Not allowing a service worker to register other service workers will encourage large, service workers that can handle all outcomes and **disfavor smaller, modular service workers** that are built to handle specific tasks. 

**It would be imprudent, inefficient, and just not cool to do so.** 

Nonetheless, security is going to be a major concern with this, so it would be best to proceed with caution and perhaps impose some limitations on service workers registering each other.

**Comment by wanderview** [comment]
I agree it would be prudent to restrict `register()` to contexts rooted by an open window.  So only permit it in worker contexts such that:

* a dedicated Worker created by a Window
* a dedicated Worker created by a parent Worker that is ultimately owned by a Window
* a SharedWorker with a Window attached

I think blocking `register()` in ServiceWorker would be reasonable for now.  We can always open it later if necessary.

**Comment by alastaircoote** [comment]
Just to chime in with some thoughts here: I have some experience using project-specific service workers on a site that also has a site-wide worker (some details [here](https://medium.com/the-guardian-mobile-innovation-lab/best-practices-for-using-notifications-and-service-workers-on-a-news-site-e90f01756c4d#7c5b)) and it would be incredibly useful for the two to be able to communicate, particularly given that they share the same storage allowance - the site-wide worker could tell old project workers to clean up storage/unregister them, etc. etc.

So, agreeing that it makes sense to limit `register()` for security reasons (though I'd like that too!), it would be great if we could still get the rest of `navigator.serviceWorker` in worker contexts, so that we can use `getRegistrations()`, then `postMessage()`, etc. etc.

**Comment by jakearchibald** [comment]
I think we'd be able to allow registering a service worker from a service worker once we tighten up our story around service worker lifetime https://github.com/w3c/ServiceWorker/issues/980#issuecomment-248920422.

But rejecting until we have that properly implemented seems fine.

**Comment by monica-ch** [comment]
[Bug triage] Keeping it open for further discussion

**Comment by asutherland** [comment]
As an update for any future discussion, Firefox did ship a fully functional `navigator.serviceWorker` ServiceWorkerContainer in all workers including ServiceWorkers, and we implemented the lifetime mechanism discussed in https://github.com/w3c/ServiceWorker/issues/980#issuecomment-248920422 with the main notes that:
- We did not implement any heuristics along the lines of https://github.com/w3c/ServiceWorker/issues/980#issuecomment-314487339 that proposed that the SW might be kept alive by the existence of controlled clients.
- We do have a check [mozilla::dom::ServiceWorkerLifetimeExtension::LifetimeExtendsIntoTheFuture](https://searchfox.org/firefox-main/rev/4eb5d723d627edec42ca3e5d606e1227c656dfca/dom/serviceworkers/ServiceWorkerLifetimeExtension.h#61-71) that we use to ensure that there is some minimum amount of lifetime left for situations where we would potentially spawn a ServiceWorker.  Currently the constant used means that there needs to be 5 seconds of lifetime left.  This is partly because we do continue to have a grace period for ServiceWorkers and the lifetime mechanism is separate from that (for now).  But the grace periods don't allow an end-run around lifetime extensions; a SW in its grace period has no lifetime which it can grant to others.

**Comment by yoshisatoyanagisawa** [comment]
Thanks for the detailed update, @asutherland. It's very interesting to see how Firefox managed to mitigate the nesting/abuse concerns using UA-internal lifetime constraints.

Some contributors on the Chromium side tried to expose `navigator.serviceWorker` in dedicated workers last year (crbug.com/40364838), but we found it extremely tedious and complex to implement correctly due to underlying threading and context-binding issues in Chromium. As a result, that effort has stalled, and we don't have active plans to ship it in the near future.

As you can see, there were discussions about potentially restricting `register()` or removing it from the spec due to these complexities and the potential for abuse. However, Firefox has shown a way to mitigate the risk without such restrictions. Now, it could be a choice to allow `register()` as-is while expecting UAs to implement proper lifetime restrictions.

Anyway, let me leave this open for further discussion.

**Comment by wanderview** [comment]
@asutherland Is there a WPT reflecting the firefox behavior?  It might be nice to capture this behavior and engine difference in the WPT output.
